### PR TITLE
Rails Capstone Project 

### PR DESCRIPTION
**Pull request by SaboorSohaib**
# Hi Dear :wave: 

**This is my capstone rails project.**
- Loom video Link(https://www.loom.com/share/afd7bd61a0f348c693d2edd64781beed)

### Interactions

**:diamonds: Splash screen**
  - A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

**:diamonds: Sign up and log in pages**
  - The user should be able to register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

**:diamonds: Home page (categories page)**
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that 
   category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

**:diamonds: Transactions page**
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in 
    that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
When the user clicks on the "Back" button (<), the user navigates to the home page.

**:diamonds: Add a new category page**
   - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
   - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
   - When the user clicks on the "Back" button (<), the user navigates to the home page.

**:diamonds: Add a new transaction page**
   - The user fills out a form to create a new transaction with:
   - name (mandatory)
   - amount (mandatory)
   - categories (mandatory at least one)
   - The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
   - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

**Thank You.**

### CODE REVIEW

File: app/models/group.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class Group < ApplicationRecord
+  belongs_to :user
+  has_many :group_expenses
+  has_many :expenses, through: :group_expenses
+  validates :name, presence: true, length: { maximum: 50 }
+  validates :icon, presence: true
+end
```
Comment (zieeco) [inline app/models/group.rb]: - Hey, great job so far. In the group, there should be an icon, however, you've added the icon but it should be an image. Currently, your app doesn't seem to show where the users can add an image. see this link on how to [add image upload in a rails-app](https://medium.com/@anaharris/how-to-add-image-upload-functionality-to-your-rails-app-9f7fc3f3d042)

*your app*
Screenshot from 2023-04-27 13-45-33

*design template*
Screenshot from 2023-04-27 13-49-00

File: app/views/expenses/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      </div>
+    </div>
+  </div>
+  <div>
+
+    <div class="ms-3 mt-3 mb-3">
+      Icon: <img src="<%= @group.icon %>", alt="group-image", style="width:60px; height:60px;">
+    </div>
+    <div class="ms-3 mb-3">
+      <h6>Total Amount: <%= @group.expenses.sum(:amount) %> $</h6>
```
Comment (zieeco) [inline app/views/expenses/index.html.erb]: - Hey :wave: the dollar sign should be added before the amount e.g $100, not 100$, as you've added. kindly improve on this.

File: app/views/expenses/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    </div>
+
+    <div class="d-flex flex-column gap-2 mb-3 ms-3 me-3">
+      <% @expenses.order(created_at: :desc).each do |expense| %>
+        <div class="d-flex flex-column border border-primary p-1">
+          <div>Transaction Name: <%= expense.name %></div>
+          <div>Transaction Amount: <%= expense.amount %>$</div>
```
Comment (zieeco) [inline app/views/expenses/index.html.erb]: - Hey, also, from the design, you don't need to specify the `Transaction Name:` and `Transaction Amount`
- Let's follow the design layout as much as we can

*your design*
image


*design template*
image

File: README.md [hunk 4]
```diff
@@ -1 +1,193 @@
...
+- It has user authentication.
+- It is well designed.
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+## 🚀 Live Demo <a name="live-demo"></a>
+
+- [https://sohaib-budget.onrender.com]
+- Please use this link for icon whenever you want to create a category(https://coffee.alexflipnote.dev/random)
```
Comment (zieeco) [inline README.md]: - Hey, this is great, but this won't improve the user experience of your app. Remember the users will interact with your live demo link and not your readme file. How do users get to know about using this particular URL to always want to add an icon in your app when they can add it from their PC locally from their files.? 
- I left a comment on how to upload images in a rails app. let's implement it to improve your app and to meet the project requirement.

Comment (zieeco) [reply README.md]: ### PRESENTATION VIDEO


- Hey, In your [video](https://www.loom.com/share/65c234b48c3648129eb87d4ba06b7ce3), you've not shown a highlight about a piece of the code you find interesting or are proud of making. let's add this missing feature in the video by recording another video.

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @SaboorSohaib,




Great job so far! 👏 
You've done well, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- No linter errors.✔️
- Descriptive PR. ✔️
- Good commit messages :+1: 
- Nice design :+1: 

## Required Changes ♻️
_Check the comments under the review._




## Optional suggestions
- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ShadyShawkat — APPROVED** [review]
## Hi @SaboorSohaib, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- All linters are passing ✔️ 
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**